### ISSUE TITLE
OLM 0.14.2 ignores subscription channel changes.

### ISSUE THREAD

**Issue by cdjohnson** [operator-framework/operator-lifecycle-manager#1510]
## Bug Report

**What did you do?**
I've created a multi-channel operator that uses the Automatic z-build strategy using a Catalog Index Image/Operator Bundle and CatalogSource.

- Channel V0.0 has CSV version 0.0.12
- Channel V1.0 has CSV 1.0.2 with a `olm.skipRange` of `>=1.0.0 <1.0.2`  

1.  Create a CatalogSource as follows:
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: panamax-operators
  namespace: openshift-marketplace
spec:
  displayName: Panamax Operators
  image: quay.io/cdjohnson/ibm-sample-panamax-catalog@sha256:e467fdb339d9c24229041a99c5ea49cea1f373230077d68c3e121cd3c03eab95
  publisher: IBM
  sourceType: grpc
```
2.  Create a namespace: `panamax`
3.  Create a Subscription to the `V0.0` channel:
```
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: ibm-sample-panamax
  namespace: panamax
spec:
  channel: V0.0
  installPlanApproval: Automatic
  name: ibm-sample-panamax
  source: panamax-operators
  sourceNamespace: openshift-marketplace
```
4.  Verify that the InstallPlan and CSV are properly installed.  I see:
```
NAME                         DISPLAY                VERSION   REPLACES   PHASE
ibm-sample-panamax.v0.0.12   IBM Panamax Operator   0.0.12               Succeeded
```

5.  Enable debug=true on the `operator-catalog` deployment
6.  Patch the subscription to channel: V1.0
    `oc patch sub ibm-sample-panamax -p '{"spec": {"channel": "V1.0"}}' --type merge`

**What did you expect to see?**
I'm expecting a new InstallPlan to be generated that will upgrade the CSV to the head version in the V1.0 channel.

**What did you see instead? Under which circumstances?**
I didn't see any change whatsoever.  From the logs:
```
$ oc -n openshift-operator-lifecycle-manager logs deploy/catalog-operator
time="2020-05-09T19:34:49Z" level=info msg=syncing event=update reconciling="*v1alpha1.Subscription" selflink=/apis/operators.coreos.com/v1alpha1/namespaces/panamax/subscriptions/ibm-sample-panamax
time="2020-05-09T19:34:49Z" level=debug msg="reconciliation successful" event=update reconciling="*v1alpha1.Subscription" selflink=/apis/operators.coreos.com/v1alpha1/namespaces/panamax/subscriptions/ibm-sample-panamax state="*subscription.catalogHealthyState"
time="2020-05-09T19:34:49Z" level=debug msg="resolving sources" id=3B1Xa namespace=panamax
time="2020-05-09T19:34:49Z" level=debug msg="checking if subscriptions need update" id=3B1Xa namespace=panamax
time="2020-05-09T19:34:49Z" level=debug msg="resolving subscriptions in namespace" id=3B1Xa namespace=panamax
```

**Environment**
* operator-lifecycle-manager version:
Using OpenShift 4.4.3 that uses OLM 0.14.2.

* Kubernetes version information:
```
Client Version: openshift-clients-4.4.0-202003060720
Server Version: 4.4.3
Kubernetes Version: v1.17.1
```
* Kubernetes cluster kind: 
OpenShift UPI on amd64 coreos.

**Possible Solution**

**Additional context**

This is what the (abbreviated) PackageManifest looks like for my catalog:
```
status:
  catalogSource: panamax-operators
  catalogSourceDisplayName: Panamax Operators
  catalogSourceNamespace: openshift-marketplace
  catalogSourcePublisher: IBM
  channels:
  - currentCSV: ibm-sample-panamax.v0.0.12
    currentCSVDesc:
      annotations: 
       capabilities: Basic Install
      version: 0.0.12
    name: V0.0
  - currentCSV: ibm-sample-panamax.v1.0.1
    currentCSVDesc:
      annotations:
        capabilities: Basic Install
        olm.skiprange: '>=1.0.0 <1.0.1'
      version: 1.0.1
    name: V1.0
```

**Comment by cdjohnson** [comment]
I've also tried using:  olm.skiprange: ‘>=0.0.0 <1.0.1’

Same problem:
Updated catalog image:
http://quay.io/cdjohnson/ibm-sample-panamax-catalog@sha256:eda7724ef96d2f8e7be2d0049c7d8b0591ffef926d38f090b91cf0ae7d645645

**Comment by shawn-hurley** [comment]
@kevinrizza @gallettilance 

This appears to be because we are not setting the `skiprange` field in the operator bundle.

```
sqlite> select name, skiprange from operatorbundle where bundle like '%skipRange%';
ibm-sample-panamax.v1.0.1|
ibm-sample-panamax.v1.1.1|
ibm-sample-panamax.v2.0.0|
```


@cdjohnson any chance you can tell us which version of OPM you were using?

**Comment by ecordell** [comment]
@cdjohnson it looks like the annotation is incorrect:

```
      "olm.skiprange": ">=1.0.0 <2.0.0"
```

This should be:

```
      "olm.skipRange": ">=1.0.0 <2.0.0"
```

Please see the [docs here](https://github.com/operator-framework/operator-lifecycle-manager/blob/0943c4fa7384678cbf2a4a171be476aa5d1327f5/doc/design/how-to-update-operators.md#skiprange) for more details. I do think that this is something that we can improve, we should probably notice things with a certain edit difference from known fields.

**Comment by cdjohnson** [comment]
Ugh.  So it looks like I'm stupid on two fronts:

1.  I definitely had the wrong annotation.  
2. I was thinking that `olm.skipRange` REPLACED the `replaces` value.

In reality, it COMPLEMENTS `replaces`.

Here's a summary that would be nice to spell out in the docs:

1. `replaces` is ALWAYS REQUIRED to build a graph.
2. Omit `replaces` to represent an origin of a New graph.
3. Every channel has its own Head.
4. Movement from a current node to Head always visits every node.
5.  Use `olm.skipRange` to skip visiting select nodes.

Can you guys confirm my understanding?  I think i had heard that I shouldn't be using `replaces`...

**Comment by cdjohnson** [comment]
When I use `replaces` AND `olm.skipRange`, I have the ability to change my channel, but the skips are completely ignored.

If I don't add `replaces` but only add `olm.skipRange`, I can see it being added to the `operatorbundle.skiprange` field now, but each channel only has a single bundle in it.  The head, and it won't let me switch channels.

Totally confused.

**Comment by cdjohnson** [comment]
I did some testing with the new bundle format using opm 1.12.2 and OpenShift 4.4.3, which uses OLM 0.14.2

My assertions are:
1.  The `replaces` attribute MUST ALWAYS refer to the previous HEAD of the graph.  
2. The `replaces` attribute MUST NEVER refer to a node that has already been replaced.  Doing so will render any nodes in-between as orphans, effectively erasing them from the graph.
3.  The Channel ONLY describes which node in the graph is the head of that particular channel.
4.  When using `olm.skipRange`, this allows OLM to skip over nodes in the graph that are included in the range.

I've created a graph of versions as follows (chronologically topdown).  The catalog image can be found [here](quay.io/cdjohnson/ibm-sample-panamax-catalog:bgtest5):
|  Version | Replaces  | SkipRange | Default Channel |  Channels | 
|---|---|---|---|---|
| 1.0.0 |  | | V1.0 | V1.0 |
| 1.0.1 | 1.0.0 | <1.0.1 | V1.0 | V1.0 |
| 1.1.0 | 1.0.1 | <1.1.0 | V1.1 | V1.1 |
| 1.1.1 | 1.1.0 | <1.1.1 | V1.1 | V1.1 |
| 1.2.0 | 1.1.1 | <1.2.0 | V1.2 | V1.2 |
| 1.0.2 | 1.2.0 | <1.0.2 >=1.1.0 | V1.0 | V1.0 |

So the DAG is in chronological order, which is required to avoid orphans.

However, when displaying the channel entries, I see 1.1.1 and 1.2.0 now in the V1.0 channel:
```
--channel-entry--
1|V1.0|ibm-sample-panamax|ibm-sample-panamax.v1.0.2|2|0
2|V1.0|ibm-sample-panamax|ibm-sample-panamax.v1.2.0|3|1  << 
3|V1.0|ibm-sample-panamax|ibm-sample-panamax.v1.1.1|4|2  << 
4|V1.0|ibm-sample-panamax|ibm-sample-panamax.v1.1.0|5|3  << 
5|V1.0|ibm-sample-panamax|ibm-sample-panamax.v1.0.1|6|4
6|V1.0|ibm-sample-panamax|ibm-sample-panamax.v1.0.0||5

7|V1.1|ibm-sample-panamax|ibm-sample-panamax.v1.1.1|8|0
8|V1.1|ibm-sample-panamax|ibm-sample-panamax.v1.1.0|9|1
9|V1.1|ibm-sample-panamax|ibm-sample-panamax.v1.0.1|10|2
10|V1.1|ibm-sample-panamax|ibm-sample-panamax.v1.0.0||3

11|V1.2|ibm-sample-panamax|ibm-sample-panamax.v1.2.0|12|0
12|V1.2|ibm-sample-panamax|ibm-sample-panamax.v1.1.1|13|1
13|V1.2|ibm-sample-panamax|ibm-sample-panamax.v1.1.0|14|2
14|V1.2|ibm-sample-panamax|ibm-sample-panamax.v1.0.1|15|3
15|V1.2|ibm-sample-panamax|ibm-sample-panamax.v1.0.0||4
```

The API also reflects this problem:
```
grpcurl -plaintext -d '{"csvName": "ibm-sample-panamax.v1.0.1"}' localhost:50051 api.Registry.GetChannelEntriesThatReplace
{
  "packageName": "ibm-sample-panamax",
  "channelName": "V1.0",
  "bundleName": "ibm-sample-panamax.v1.1.0",
  "replaces": "ibm-sample-panamax.v1.0.1"
}
{
  "packageName": "ibm-sample-panamax",
  "channelName": "V1.1",
  "bundleName": "ibm-sample-panamax.v1.1.0",
  "replaces": "ibm-sample-panamax.v1.0.1"
}
{
  "packageName": "ibm-sample-panamax",
  "channelName": "V1.2",
  "bundleName": "ibm-sample-panamax.v1.1.0",
  "replaces": "ibm-sample-panamax.v1.0.1"
}

```

Likewise, when creating a subscription in OLM with channel V1.0 and startingCSV of 1.0.1, it did indeed install version 1.1.1, though it should have skipped to 1.0.2.

Attached are the CSV files I've used to build this catalog.
[bgtest5.tar.gz](https://github.com/operator-framework/operator-lifecycle-manager/files/4640584/bgtest5.tar.gz)


In Summary:  I need to understand how one would use the `replaces` and `olm.skipRange` to build proper X.Y channels using the bundle format.

**Comment by ecordell** [comment]
> 1. The replaces attribute MUST ALWAYS refer to the previous HEAD of the graph.

I would not say "MUST ALWAYS" but "SHOULD" - you can intentionally orphan a piece of the graph if you so choose.

> 2. The replaces attribute MUST NEVER refer to a node that has already been replaced. Doing so will render any nodes in-between as orphans, effectively erasing them from the graph.

Correct, you cannot have two answers for "what replaces X" in a single package/channel. 

> 3. The Channel ONLY describes which node in the graph is the head of that particular channel.

Correct, see [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/how-to-update-operators.md#replaces--channels) for more docs.

> 4. When using olm.skipRange, this allows OLM to skip over nodes in the graph that are included in the range.

Also correct. It is something of a break-glass feature. The feature that works more like you are probably expecting is the `--mode=semver-skippatch` for `opm index add`.


> However, when displaying the channel entries, I see 1.1.1 and 1.2.0 now in the V1.0 channel:

This looks correct to me, based on the table you're showing (using -> for "replaces")

channel v1.0 head: 1.0.2
1.0.2 -> 1.2.0 
1.2.0 -> 1.1.1 
1.1.1 -> 1.1.0

Which matches the rules you correctly identified above. If I had to single out an error, I would expect 1.0.2 to replace 1.0.0, not 1.2.0?

>  I need to understand how one would use the replaces and olm.skipRange to build proper X.Y channels using the bundle format.

To clarify, I assume the behavior you're looking for is:

- When subscribing to X.Y for the first time, should install the latest Z
- When switching from X.Y channel to X.Y+1, should install the latest Z
- When multiple updates are available to the latest X.Y.Z in an X.Y channel, skip to the latest Z

It looks like you want 3 channels: 1.0, 1.1, 1.2

And you want to be able to switch to X.Y+1 and get updates?

v1.0: 1.0.2 (skipRange: > 1.0.0 < 1.0.2) -> 1.0.1 (skipRange > 1.0.0 < 1.0.1) -> 1.0.0 
v1.1: 1.1.1 (skipRange: > 1.0.0 < 1.1.1) -> 1.1.0
v1.2: 1.2.1 (skipRange: > 1.1.0 < 1.2.1) -> 1.2.0

When you publish a new Z release to an X.Y channel, have the skipRange be >X.Y-1 and < X.Y.Z (the current). 

It's also possible to completely omit the "replaces" if you're using this mechanism, since the skipRange mechanism will always pick the latest Z.

**Comment by ecordell** [comment]
> when creating a subscription in OLM with channel V1.0 and startingCSV of 1.0.1, it did indeed install version 1.1.1, though it should have skipped to 1.0.2.

The skipRange you have for 1.0.2 is:

<1.0.2 >=1.1.0 

which looks backwards to me. Anything that is <1.0.2 cannot be >=1.1.0 in semver order.

**Comment by cdjohnson** [comment]
Thanks for the help.  I was able to get my scenario working if I did the following:

v1.0: 1.0.2 (skipRange: >= 1.0.0 < 1.0.2) -> 1.0.1 (skipRange >= 1.0.0 < 1.0.1) -> 1.0.0
v1.1: 1.1.1 (skipRange: >= 1.0.0 < 1.1.1) -> 1.1.0 (skipRange: >= 1.0.0 < 1.1.0)
v1.2: 1.2.1 (skipRange: >= 1.1.0 < 1.2.1) -> 1.2.0 (skipRange: >= 1.1.0 <= 1.2.0)

I had to use a combination of replaces for patch updates in a channel and skipRange for patch updates AND cross-channel upgrades.  I "think" the replaces values are optional if the skipRange is there, but I do want to be able to allow customers to select an startingCSV if needed to a patch level, which can aid in debugging.

I also had to keep the lower bound of the skipRange the same for all nodes in a channel.
